dns_server/src/server/mod.rs

```diff
@@ -0,0 +1,286 @@
+const SOA_RETRY: i32 = 86400;
+const SOA_EXPIRE: i32 = 2592000;
+const SOA_MINIMUM: u32 = 604800;
+
+// Same values as in `https://github.com/sipa/bitcoin-seeder/blob/master/dns.cpp`
+const TTL_IP: u32 = 3600;
+const TTL_NS: u32 = 21600;
+const TTL_SOA: u32 = 21600;
+
+// Maximum number of IPv4 addresses in result
+const MAX_IPV4_RECORDS: usize = 24;
+
+// Maximum number of IPv6 addresses in result
+const MAX_IPV6_RECORDS: usize = 14;
+
+impl Server {
+    pub async fn new(
+        config: Arc<DnsServerConfig>,
+        command_tx: mpsc::UnboundedReceiver<ServerCommands>,
+    ) -> Result<Self, DnsServerError> {
+        let host = Name::from_str(&config.host)?;
+        let ns = Name::from_str(&config.ns)?;
+        let mbox = Name::from_str(&config.mbox)?;
+
+        let inner = InMemoryAuthority::empty(host.clone(), ZoneType::Primary, false);
+
+        let auth = Arc::new(AuthorityImpl {
+            serial: Default::default(),
+            host,
+            ns,
+            mbox,
+            inner,
+            ip4: Default::default(),
+            ip6: Default::default(),
+        });
+
+        let mut catalog = Catalog::new();
+
+        catalog.upsert(
+            LowerName::from_str(&config.host)?,
+            Box::new(Arc::clone(&auth)),
+        );
+
+        let mut server = ServerFuture::new(catalog);
+
+        for bind_addr in config.bind_addr.iter() {
+            let socket_addr: SocketAddr = bind_addr.parse()?;
+            let udp_socket = UdpSocket::bind(socket_addr).await?;
+            server.register_socket(udp_socket);
+        }
+
+        Ok(Self {
+            auth,
+            server,
+            command_tx,
+        })
+    }
+
+    pub async fn run(self) -> Result<void::Void, DnsServerError> {
+        let Server {
+            auth,
+            server,
+            mut command_tx,
+        } = self;
+
+        tokio::spawn(async move {
+            while let Some(command) = command_tx.recv().await {
+                handle_command(&auth, command);
+            }
+        });
+
+        server.block_until_done().await?;
+
+        Err(DnsServerError::Other(
+            "trust_dns_server terminated unexpectedly",
+        ))
+    }
+}
+
+/// Wrapper for InMemoryAuthority that selects random addresses every second
+struct AuthorityImpl {
+    serial: AtomicU32,
+    host: Name,
+    ns: Name,
+    mbox: Name,
+    inner: InMemoryAuthority,
+    ip4: Mutex<Vec<Ipv4Addr>>,
+    ip6: Mutex<Vec<Ipv6Addr>>,
+}
+
+impl AuthorityImpl {
+    async fn refresh(&self) {
+        let new_serial = std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
+            .expect("valid time expected")
+            .as_secs() as u32;
+        let old_serial = self.serial.swap(new_serial, Ordering::Relaxed);
+        if old_serial == new_serial {
+            return;
+        }
+
+        let ipv4 = self
+            .ip4
+            .lock()
+            .expect("mutex must be valid")
```

Comment:
Perhaps a dumb comment... but please change "mutex must be valid" to "mutex must be valid (ipv4)". Using the same message in multiple places can lead to confusion when an error happens.

Reply:
Sure, updated.